File: src/engine/sprite-engine.lua
local Camera = require('engine/camera')

---@class SpriteEngine
---@field camera Camera
---@field actors Actor[]
SpriteEngine = {}
SpriteEngine.__index = SpriteEngine

function SpriteEngine:new(x, y)
    local obj = {
        x = x,
        y = y,
        actors = {},
        camera = Camera:new(0, 0, 1, 200)
    }
    setmetatable(obj, SpriteEngine)
    return obj
end

-- Add Actor method
function SpriteEngine:addActor(actor)
    table.insert(self.actors, actor)
end

-- Remove Actor method
function SpriteEngine:removeActor(actor)
    for i, a in ipairs(self.actors) do
        if a == actor then
            table.remove(self.actors, i)
            break
        end
    end
end

-- Update method
function SpriteEngine:update(dt)
    if love.keyboard.isDown('kp+') then
        self.camera.scale = self.camera.scale - self.camera.speed * dt * 0.001
    elseif love.keyboard.isDown('kp-') then
        self.camera.scale = self.camera.scale + self.camera.speed * dt * 0.001
    end

    for _, actor in ipairs(self.actors) do
        actor:update(dt)
    end
end

-- Draw method
function SpriteEngine:draw()
    self.camera:set()
    for _, actor in ipairs(self.actors) do
        actor:draw()
    end
    self.camera:unset()
end

return SpriteEngine

File: src/engine/sprite-engine.ts
import { Screen } from "love.graphics";
import { Actor } from "./actor";
import { Camera } from "./camera";

export class SpriteEngine {
  private actors: Actor[] = [];

  constructor(public camera: Camera = new Camera()) {
    this.actors = [];
  }

  getActors() {
    return [...this.actors];
  }

  addActor(actor: Actor) {
    this.actors.push(actor);
  }

  removeActor(actor: Actor) {
    this.actors = this.actors.filter((a) => a !== actor);
  }

  update(dt: number) {
    if (love.keyboard.isDown && love.keyboard.isDown("kp+")) {
      this.camera.scale -= this.camera.speed * dt * 0.001;
    } else if (love.keyboard.isDown && love.keyboard.isDown("kp-")) {
      this.camera.scale += this.camera.speed * dt * 0.001;
    }
    for (const actor of this.actors) {
      actor.update(dt);
    }
  }

  draw(screen?: Screen) {
    this.camera.set();
    for (const actor of this.actors) {
      actor.draw(screen);
    }
    this.camera.unset();
  }
}
